File: python/agentic_memory/memory.py
"""AgentMemory - the main entry point."""

from __future__ import annotations

from typing import Any, Optional

from .types import (
    MemoryEntry,
    MemoryType,
    ImportanceLevel,
    RetrievalQuery,
    RetrievalResult,
    ConflictResult,
    Checkpoint,
    TaskNode,
    DecayConfig,
)
from .utils import generate_id, now
from .store.local import LocalStore
from .retrieval.embedder import BuiltinEmbedder
from .retrieval.multi_signal import MultiSignalRetriever
from .retrieval.conflict import ConflictDetector
from .decay import DecayEngine


class AgentMemory:
    """
    Framework-agnostic memory layer for AI agents.

    Example::

        memory = AgentMemory()

        await memory.store(
            content="User prefers dark mode",
            type=MemoryType.PREFERENCE,
            scope="user:123",
        )

        results = await memory.retrieve(RetrievalQuery(query="UI preferences"))
        conflicts = await memory.check_conflicts("User prefers light mode", "user:123")
    """

    def __init__(
        self,
        *,
        store: Any = None,
        embedder: Any = None,
        decay: DecayConfig | None = None,
        default_scope: str = "default",
        max_checkpoints: int = 10,
    ) -> None:
        self._backend = store if store is not None else LocalStore()
        self._embedder = embedder if embedder is not None else BuiltinEmbedder()
        self._retriever = MultiSignalRetriever(self._backend, self._embedder)
        self._conflict_detector = ConflictDetector(self._backend, self._embedder)
        self._decay_engine = DecayEngine(decay)
        self._checkpoints: dict[str, Checkpoint] = {}
        self._default_scope = default_scope
        self._max_checkpoints = max_checkpoints

    # ─── Store ───

    async def store(
        self,
        content: str,
        *,
        type: MemoryType = MemoryType.FACT,
        scope: Optional[str] = None,
        importance: ImportanceLevel = ImportanceLevel.SOFT,
        confidence: float = 1.0,
        metadata: Optional[dict[str, Any]] = None,
    ) -> MemoryEntry:
        """Store a new memory entry."""
        if not content or not content.strip():
            raise ValueError("Memory content cannot be empty")

        # Train before embed so IDF stats include this document
        if isinstance(self._embedder, BuiltinEmbedder):
            self._embedder.train([content])

        embedding = await self._embedder.embed(content)

        entry = MemoryEntry(
            id=generate_id(),
            content=content,
            type=type,
            scope=scope or self._default_scope,
            importance=importance,
            confidence=confidence,
            embedding=embedding,
            metadata=metadata or {},
            created_at=now(),
            updated_at=now(),
            version=1,
        )

        await self._backend.set(entry)
        return entry

    async def update(
        self,
        id: str,
        *,
        content: Optional[str] = None,
        type: Optional[MemoryType] = None,
        importance: Optional[ImportanceLevel] = None,
        confidence: Optional[float] = None,
        metadata: Optional[dict[str, Any]] = None,
    ) -> Optional[MemoryEntry]:
        """Update an existing memory entry."""
        entry = await self._backend.get(id)
        if entry is None:
            return None

        if content is not None:
            if not content.strip():
                raise ValueError("Memory content cannot be empty")
            entry.content = content

        if type is not None:
            entry.type = type
        if importance is not None:
            entry.importance = importance
        if confidence is not None:
            entry.confidence = confidence
        if metadata is not None:
            entry.metadata = metadata

        entry.updated_at = now()
        entry.version += 1

        # Re-embed if content changed
        if content is not None:
            if isinstance(self._embedder, BuiltinEmbedder):
                self._embedder.train([content])
            entry.embedding = await self._embedder.embed(content)

        await self._backend.set(entry)
        return entry

    async def get(self, id: str) -> Optional[MemoryEntry]:
        """Get a memory by ID."""
        return await self._backend.get(id)

    async def delete(self, id: str) -> bool:
        """Delete a memory by ID."""
        return await self._backend.delete(id)

    async def get_all(self, scope: Optional[str] = None) -> list[MemoryEntry]:
        """Get all memories, optionally filtered by scope."""
        return await self._backend.get_all(scope)

    async def clear(self, scope: Optional[str] = None) -> None:
        """Clear all memories, optionally for a specific scope."""
        return await self._backend.clear(scope)

    # ─── Retrieve ───

    async def retrieve(self, query: RetrievalQuery) -> list[RetrievalResult]:
        """Multi-signal retrieval."""
        return await self._retriever.retrieve(query)

    # ─── Conflict Detection ───

    async def check_conflicts(
        self, content: str, scope: Optional[str] = None
    ) -> list[ConflictResult]:
        """Check if content conflicts with stored memories."""
        return await self._conflict_detector.check(
            content, scope or self._default_scope
        )

    # ─── Decay ───

    def get_decayed_confidence(self, entry: MemoryEntry) -> float:
        """Get the current effective confidence after decay."""
        return self._decay_engine.compute_decayed_confidence(entry)

    async def cleanup(
        self, scope: Optional[str] = None, threshold: float = 0.01
    ) -> int:
        """Clean up expired memories. Returns count deleted."""
        entries = await self._backend.get_all(scope)
        expired = self._decay_engine.get_expired(entries, threshold)
        for entry in expired:
            await self._backend.delete(entry.id)
        return len(expired)

    # ─── Checkpointing ───

    async def checkpoint(
        self,
        *,
        task_graph: list[TaskNode],
        summary: str,
        tool_outputs: Optional[dict[str, Any]] = None,
        active_memory_ids: Optional[list[str]] = None,
    ) -> Checkpoint:
        """Create a checkpoint of current task state."""
        cp = Checkpoint(
            id=generate_id(),
            task_graph=task_graph,
            summary=summary,
            tool_outputs=tool_outputs or {},
            active_memory_ids=active_memory_ids or [],
            created_at=now(),
        )

        self._checkpoints[cp.id] = cp

        # Enforce max checkpoints (LRU eviction)
        if len(self._checkpoints) > self._max_checkpoints:
            oldest_key = next(iter(self._checkpoints))
            del self._checkpoints[oldest_key]

        return cp

    async def rehydrate(
        self, checkpoint_id: str
    ) -> Optional[dict[str, Any]]:
        """Rehydrate from a checkpoint. Returns checkpoint + memories."""
        cp = self._checkpoints.get(checkpoint_id)
        if cp is None:
            return None

        memories: list[MemoryEntry] = []
        for mid in cp.active_memory_ids:
            entry = await self._backend.get(mid)
            if entry is not None:
                memories.append(entry)

        return {"checkpoint": cp, "memories": memories}

    def get_latest_checkpoint(self) -> Optional[Checkpoint]:
        """Get the latest checkpoint."""
        if not self._checkpoints:
            return None
        return list(self._checkpoints.values())[-1]

    def list_checkpoints(self) -> list[Checkpoint]:
        """List all checkpoints."""
        return list(self._checkpoints.values())

File: src/core/memory.ts
import type {
  AgentMemoryConfig,
  MemoryEntry,
  MemoryType,
  ImportanceLevel,
  RetrievalQuery,
  RetrievalResult,
  ConflictResult,
  StorageBackend,
  Embedder,
  Checkpoint,
  TaskNode,
} from './types.js';
import { generateId, now } from './utils.js';
import { LocalStore } from '../store/local.js';
import { BuiltinEmbedder } from '../retrieval/embedder.js';
import { MultiSignalRetriever } from '../retrieval/multi-signal.js';
import { ConflictDetector } from '../retrieval/conflict.js';
import { DecayEngine } from './decay.js';

/**
 * AgentMemory - the main entry point.
 *
 * Framework-agnostic memory layer for AI agents.
 * Handles storage, multi-signal retrieval, conflict detection,
 * typed decay, and checkpointing.
 *
 * @example
 * ```typescript
 * const memory = new AgentMemory();
 *
 * await memory.store({
 *   content: 'User prefers dark mode',
 *   type: 'preference',
 *   scope: 'user:123',
 * });
 *
 * const results = await memory.retrieve({ query: 'UI preferences' });
 * const conflicts = await memory.checkConflicts('User prefers light mode', 'user:123');
 * ```
 */
export class AgentMemory {
  private backend: StorageBackend;
  private embedder: Embedder;
  private retriever: MultiSignalRetriever;
  private conflictDetector: ConflictDetector;
  private decayEngine: DecayEngine;
  private checkpoints: Map<string, Checkpoint> = new Map();
  private defaultScope: string;
  private maxCheckpoints: number;

  constructor(config: AgentMemoryConfig = {}) {
    // Initialize storage backend
    this.backend = config.store === 'local' || !config.store
      ? new LocalStore()
      : config.store;

    // Initialize embedder
    this.embedder = config.embedder === 'builtin' || !config.embedder
      ? new BuiltinEmbedder()
      : config.embedder;

    // Initialize retriever and conflict detector
    this.retriever = new MultiSignalRetriever(this.backend, this.embedder);
    this.conflictDetector = new ConflictDetector(this.backend, this.embedder);
    this.decayEngine = new DecayEngine(config.decay);

    this.defaultScope = config.defaultScope ?? 'default';
    this.maxCheckpoints = config.checkpoint?.maxCheckpoints ?? 10;
  }

  // ─── Store ───

  /**
   * Store a new memory entry.
   * Automatically generates ID, timestamps, and embedding.
   */
  async store(params: {
    content: string;
    type?: MemoryType;
    scope?: string;
    importance?: ImportanceLevel;
    confidence?: number;
    metadata?: Record<string, unknown>;
  }): Promise<MemoryEntry> {
    if (!params.content || params.content.trim().length === 0) {
      throw new Error('Memory content cannot be empty');
    }

    // Train the builtin embedder BEFORE embedding so IDF stats include this document
    if (this.embedder instanceof BuiltinEmbedder) {
      this.embedder.train([params.content]);
    }

    const embedding = await this.embedder.embed(params.content);

    const entry: MemoryEntry = {
      id: generateId(),
      content: params.content,
      type: params.type ?? 'fact',
      scope: params.scope ?? this.defaultScope,
      importance: params.importance ?? 'soft',
      confidence: params.confidence ?? 1.0,
      embedding,
      metadata: params.metadata ?? {},
      createdAt: now(),
      updatedAt: now(),
      version: 1,
    };

    await this.backend.set(entry);
    return entry;
  }

  /**
   * Update an existing memory entry.
   * Increments version and updates timestamp.
   */
  async update(id: string, updates: Partial<Pick<MemoryEntry, 'content' | 'type' | 'importance' | 'confidence' | 'metadata'>>): Promise<MemoryEntry | null> {
    const entry = await this.backend.get(id);
    if (!entry) return null;

    const updated: MemoryEntry = {
      ...entry,
      ...updates,
      updatedAt: now(),
      version: entry.version + 1,
    };

    // Re-embed if content changed
    if (updates.content && updates.content !== entry.content) {
      if (updates.content.trim().length === 0) {
        throw new Error('Memory content cannot be empty');
      }
      if (this.embedder instanceof BuiltinEmbedder) {
        this.embedder.train([updates.content]);
      }
      updated.embedding = await this.embedder.embed(updates.content);
    }

    await this.backend.set(updated);
    return updated;
  }

  /** Get a memory by ID */
  async get(id: string): Promise<MemoryEntry | null> {
    return this.backend.get(id);
  }

  /** Delete a memory by ID */
  async delete(id: string): Promise<boolean> {
    return this.backend.delete(id);
  }

  /** Get all memories, optionally filtered by scope */
  async getAll(scope?: string): Promise<MemoryEntry[]> {
    return this.backend.getAll(scope);
  }

  /** Clear all memories, optionally for a specific scope only */
  async clear(scope?: string): Promise<void> {
    return this.backend.clear(scope);
  }

  // ─── Retrieve ───

  /**
   * Multi-signal retrieval.
   * Combines similarity, recency, importance, and task-relevance.
   */
  async retrieve(query: RetrievalQuery): Promise<RetrievalResult[]> {
    return this.retriever.retrieve(query);
  }

  // ─── Conflict Detection ───

  /**
   * Check if content conflicts with stored memories.
   * Returns conflicts sorted by confidence.
   *
   * @example
   * ```typescript
   * const conflicts = await memory.checkConflicts('User likes meat', 'user:123');
   * if (conflicts.length > 0 && conflicts[0].action === 'clarify') {
   *   // Ask user to confirm preference change
   * }
   * ```
   */
  async checkConflicts(content: string, scope?: string): Promise<ConflictResult[]> {
    return this.conflictDetector.check(content, scope ?? this.defaultScope);
  }

  // ─── Decay ───

  /**
   * Get the current effective confidence of a memory after decay.
   */
  getDecayedConfidence(entry: MemoryEntry): number {
    return this.decayEngine.computeDecayedConfidence(entry);
  }

  /**
   * Clean up expired memories (decayed below threshold).
   * Returns the number of deleted entries.
   */
  async cleanup(scope?: string, threshold = 0.01): Promise<number> {
    const entries = await this.backend.getAll(scope);
    const expired = this.decayEngine.getExpired(entries, threshold);
    for (const entry of expired) {
      await this.backend.delete(entry.id);
    }
    return expired.length;
  }

  // ─── Checkpointing ───

  /**
   * Create a checkpoint of current task state.
   * Use this before context overflow to preserve state.
   */
  async checkpoint(params: {
    taskGraph: TaskNode[];
    summary: string;
    toolOutputs?: Record<string, unknown>;
    activeMemoryIds?: string[];
  }): Promise<Checkpoint> {
    const cp: Checkpoint = {
      id: generateId(),
      taskGraph: params.taskGraph,
      summary: params.summary,
      toolOutputs: params.toolOutputs ?? {},
      activeMemoryIds: params.activeMemoryIds ?? [],
      createdAt: now(),
    };

    this.checkpoints.set(cp.id, cp);

    // Enforce max checkpoints (LRU eviction)
    if (this.checkpoints.size > this.maxCheckpoints) {
      const oldest = this.checkpoints.keys().next().value;
      if (oldest) this.checkpoints.delete(oldest);
    }

    return cp;
  }

  /**
   * Rehydrate from a checkpoint.
   * Returns the checkpoint data + relevant memories.
   */
  async rehydrate(checkpointId: string): Promise<{
    checkpoint: Checkpoint;
    memories: MemoryEntry[];
  } | null> {
    const cp = this.checkpoints.get(checkpointId);
    if (!cp) return null;

    // Fetch the memories that were active at checkpoint time
    const memories: MemoryEntry[] = [];
    for (const id of cp.activeMemoryIds) {
      const entry = await this.backend.get(id);
      if (entry) memories.push(entry);
    }

    return { checkpoint: cp, memories };
  }

  /** Get the latest checkpoint */
  getLatestCheckpoint(): Checkpoint | null {
    const all = Array.from(this.checkpoints.values());
    if (all.length === 0) return null;
    return all[all.length - 1];
  }

  /** List all checkpoints */
  listCheckpoints(): Checkpoint[] {
    return Array.from(this.checkpoints.values());
  }
}
